### ISSUE TITLE
Comparing Reflection objects. We either need a story or decide we don't have a story. 

### ISSUE THREAD

**Issue by AtsushiKan** [dotnet/corefx#8130]
This came up as an underlying issue in Issue #3657. Opening this as a Future issue to track.

What's a developer to do if you need to compare two Reflection objects such as ParameterInfo for "semantic equality"?

On .NET Native, we actually did override Equals/GetHashCode on these objects and they perform semantic equality. However, .NET Native uses a completely different implementation of Reflection.

On CoreClr, Reflection objects inherit the Object Equals() method, so those checks are just reference equality checks. They don't check semantic equality, but they often fool people into thinking they do because the CLR often caches Reflection objects for reuse. We probably don't want to make the caching 100% (unnecessary memory usage) but overriding Equals/GetHashCode may also have its costs (surface area compat? What about user-extended Reflection types if we ever support that?)

**Comment by Mike-EEE** [comment]
Thank you for creating this issue @AtsushiKan!  From my perspective, it seems very intuitive/expected that these objects are essentially singleton and should preserve their equality as such.  The issue you mentioned exposes a race condition when more than one thread is used in an application for initialization.  My suspicion is that most developers assume that these objects are singletons, and the fact that System.Composition was designed in such a way reflects this assumption.

As more and more applications are on-boarded to multi-threaded scenarios, I believe we will see more instances of this issue exposed.  It would be great to see an official guarantee of some sort to align single-threaded expectations with multi-threaded expectations.

Just my two cents, however.  Keep in mind that I am a newb in this area and do not know much about impact/reach of how such a decision/guarantee would have, but hopefully sharing my perspective will assist in some way. :)

Thanks again for creating this issue.  And for taking the time to explain the issue in perfectly accurate detail.  It is much appreciated!

**Comment by SamuelEnglard** [comment]
I was dealing with this today! Needed a way to uniquely identify/compare `MethodInfo`s. What I decided t do was to use their `ToString()` since it returns the full signature of the method, including defining object. Technically doesn't include assembly but that could be easily fixed.

**Comment by Mike-EEE** [comment]
Yes strings are definitely a good fallback for the current paradigm, but I was under the impression that they are slow?  To me, it would be ideal that every `MemberInfo` has a `Guid` and that `Guid` is the same for that `MemberInfo` regardless of the thread it is on.  But that's just me. :smile:

**Comment by GSPP** [comment]
If the decision is to not guarantee reference equality it would be useful to intentionally break it. That way developers will at least find the bug during testing.

This could be done like this: `if (coinToss(probability: 1e-3)) dontUseCache = true;`.

This piece of code has many problems but maybe something better can be found that at least alerts the developer who is mistakenly relying on this.

Alternative: remember the thread ID the cache was created on. The first time another thread accesses the cache don't use the cache. That way there is exactly one obvious uniqueness breakage per application run which hopefully serves as an alert. There would be no sustained performance cost.

**Comment by Mike-EEE** [comment]
That is a good idea as well @GSPP.  From my take, to me (and others) it seems/feels like the type system _should_ be static/cached per app domain.  It is natural to expect (or assume, as what has gotten us into this issue to begin with :smile: ) that each type object is indeed the one representation for that object (or "semantically equal" as @AtsushiKan much more elegantly/efficiently put it!).  In fact, it is, it's just the silly thread racing that is keeping it from doing this.

I guess my question is... is there a scenario you can think of where the type system _shouldn't_ be static/cached per AppDomain?

**Comment by GSPP** [comment]
Some code enumerates all types and all members of all loaded assemblies. It does that once to find a few important items that it cares about. If the cache was as proposed this would leave a bug amount of waste forever.

Or, all cache contents must be held through weak references. But they are (I think) costly. Also, the GC would then regularly kill the cache contents. Weak ref based caches are bad design because object lifetime is managed by a component that has no idea about what is right.

**Comment by Mike-EEE** [comment]
well that is a good point.  Right now all of those members and types are indeed kept in memory.  There is no way currently of releasing those members from memory after they are loaded, correct?  So we might me able to improve upon that as well.  So... `MemberInfo.Guid` or `MemberInfo.GetHashCode` override then. :smile:

**Comment by GSPP** [comment]
Since it was not mentioned before: I think all of these classes should have their equality operators overloaded. That way a lot of now broken code starts to become correct after a recompile. I can't come up with any potential code that would be broken by that. This would mitigate some compatibility concerns as well.

**Comment by gulbanana** [comment]
oh dear. I'm glad this issue has been spotted, since it reminds me of a ton of code I've written that does stuff like "if (typeof(T) == expectedType)"..  is that a bug even on .net full?

**Comment by AtsushiKan** [comment]
For Types, comparing it this way is fine – this is doc’d in MSDN and so much existing code depends on it, we’d be crazy to break it even if we wanted to.

For the other MemberInfo objects (methods, fields) – there’s no official promise that I know of but realistically, it’s hard to come up with a case where this doesn’t work in the desktop and there’s probably too much compat constraints to break that either.

I’d say my suggestion would be:
-   Use Equals or operator== (which redirects you to Equals() and we’re putting operator== back into the .NetCore api) That approach lets us relax the internal caching without breaking the logic.

-   Avoid using object.ReferenceEqual() or “(object)m1 == (object)m2”. That we’re more likely to push back against.

**Comment by dotnetjunkie** [comment]
I think the memory overhead argument (as proposed by @GSPP)is nin sense, since both Type instances already caches MemberInfo instances and MemberInfo already cache ParameterInfo. So there is no extra cost in guaranteeing singletons. The only problem is a concurrency problem.

**Comment by dotnetjunkie** [comment]
The MSDN documentation is clear about `Type` instances being unique. I noticed the creation of `MemberInto` objects to be protected by a lock in the .NET 4.5 code base.

What I would like to know is whether this safety net for `MemberInfo` exists in all platforms (.NET 4.x, .NET Core, .NET Native, etc) or whether we should we should assume that `MemberInfo` instances are not unique (and don't override Equals/GetHashCode).

**Comment by GSPP** [comment]
@dotnetjunkie this might be a weakly referenced cache. I believe it was in an earlier .NET version. Not sure what it's now.

**Comment by GSPP** [comment]
@AtsushiKan what's the resolution to this issue?

**Comment by AtsushiKan** [comment]
What I said above.

Use Equals or operator== (which redirects you to Equals() and we’re putting operator== back into the .NetCore api)

Don't object.ReferenceEqual() or “(object)m1 == (object)m2”. (other than for System.Type)
